Generate fresh symbols for compound literal expressions

fbrausse opened the pull request:
Fixes #591.

The set of attributes for the new symbol may not be complete; it's hard for me to tell without documentation. I'd be happy about any suggestions how to improve the generated symbols. In particular, the `id` is still `clang_c_convertert::cl$1` and not something similar to `c:main.c@54@F@main@z` like the other local symbols and I don't know how to obtain such a string. That's why this PR is still a draft. ~I may also attempt a fix of #709 in this PR later.~

*Edit* It turns out that #709 is about "cast to union", which is handled differently from compound literals in ESBMC due to this  PR, thus we'll fix it in another attempt later. There is no need to introduce a new (read-only) symbol for casts to union, thus the bug remains. Probably somewhere in the dereference code failing to handle `constant_union2t` expressions containing a dereference.

fbrausse committed:
[clang-c] typo in comment

fbrausse committed:
extract generating temporary symbols from goto-convert to util

fbrausse committed:
[clang-c-frontend] generate symbols for compound literals, fixes #591

fbrausse committed:
[regression] add tests for #591

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
+    cl.mode = "C";
+    cl.module = get_modulename_from_path(location.file().as_string());
+    cl.location = location;
+    cl.type = t;
```
here you're probably want to use `get_decl_name` + `get_default_symbol` to generate the symbol.

You'll probably need to extend `get_decl_name` for compound literals, similar to how it's done for unnamed fields.

fbrausse replied:
I've tried to, but failed, since there does not seem to be any `clang::NamedDecl` (or any `clang::Decl`) available at this point.

I've noticed several assumptions being made on the `id` string elsewhere; is it's format defined in ESBMC? For instance, the `c:main.c@54@F@main@z` id above is purely generated by clang, yet there are many instances in which the id is compared to specific constants (mostly for intrinsics and functions in c2goto/library).

fbrausse replied:
> Not really, we use it as defined by clang.

Since there is no `clang::Decl` for this anonymous object, I don't see a way to use Clang to generate such an id for us in a meaningful way. The situation for unnamed field members is different in that they are already part of a declaration whereas the anonymous symbol we introduce does not have declaration in the AST.

The comment above the code you cited mentions `get_default_symbol()`. I'm not calling this function, because that would require some awkward re-conversion of `name`/`id` to `std::string` and back to `irep_idt` not changing any of the actual values.

mikhailramalho replied:
> The comment above the code you cited mentions `get_default_symbol()`. I'm not calling this function, because that would require some awkward re-conversion of `name`/`id` to `std::string` and back to `irep_idt` not changing any of the actual values.

Do you mean inside the new class you created? My concern is the duplicated code, as this is basically the code inside `get_default_symbol()`: the goal of this function is to create all symbols in the frontend, so we don't introduce inconsistencies by duplicating the code in other places.

fbrausse replied:
> Do you mean inside the new class you created?

I meant this comment directly above: https://github.com/esbmc/esbmc/blob/4b1d1a8e338edc65cd6c7ca58abc3f4a84445ad0/src/clang-c-frontend/clang_c_convert.cpp#L1470-L1471

> My concern is the duplicated code [...]

Right. I've changed the signature of get_default_symbol() to take `irep_idt`s directly, thus avoiding the awkward conversions. The code in question now uses this function.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
+
+    new_expr = symbol_expr(cl);
+
+    if(current_block)
```
I'm not sure about this piece of code here, maybe this can be converted to a codet, using `convert_expression_to_code` instead?

fbrausse replied:
This piece of code is inserting a declaration of the new symbol into the current block before the statement the currently converted expression belongs to. It basically follows `get_var()` here: https://github.com/esbmc/esbmc/blob/038c09a042e6d86d5682bee9cf8f20c1300bb9ef/src/clang-c-frontend/clang_c_convert.cpp#L481-L493 As discussed in https://github.com/esbmc/esbmc/pull/644#discussion_r827917073, local symbols (or those with automatic storage) are required to be declared in order for the value-set analysis to recognize them as being local to a function.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
+    }
+    else
+    {
+      /* XXX fbrausse: do we need to record the declaration on the top-level? */
```
Do we have a test case for that? I suggest we abort with a message.

fbrausse replied:
Yes, I've addded this test case:
```c
int *p = &(int){5};

int main()
{
	assert(p);
	assert(*p == 5);
	++*p;
	assert(*p == 6);
}
```

fbrausse replied:
Strange, with `--symex-trace`, `p` is assigned twice, even though the result is `VERIFICATION SUCCESSFUL`:
```
[...]
        p=&cl$1;

        cl$1=5;

        __ESBMC_num_total_threads=0;

        __ESBMC_num_threads_running=0;

        p=0;
[...]
```

mikhailramalho replied:
maybe because p is global, we default initialize it to zero?

fbrausse replied:
Sorry, my mistake, `p=0` came from strtok():
```
        // 344 file string.c line 233 function strtok
        p=0;
```
It is a symbol with different id. This code path for the file scope does not appear to require a declaration.

fbrausse committed:
[regression] add test case for compound literal on file scope

mikhailramalho commented:
Not really, we use it as defined by clang.

Em seg., 23 de mai. de 2022 às 10:55, Franz Brauße ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/clang-c-frontend/clang_c_convert.cpp
> <https://github.com/esbmc/esbmc/pull/711#discussion_r879259814>:
>
> > +     *   memcpy(p, &x, sizeof(int));
> +     *   assert(*p == 42);
> +     *
> +     * It has the same semantics as an non-anonymous variable of the
> +     * corresponding type. Thus, we introduce a new symbol to represent it.
> +     */
> +
> +    /* "cl" stands for "compound literal", but that is too long for possibly
> +     * frequent copies of the symbol's name and id. */
> +    symbolt &cl = anon_symbol.new_symbol(context, initializer.type(), "cl$");
> +    /* mimic get_default_symbol() w/o setting id and name as that's already
> +     * taken care of by anon_symbol.new_symbol() above */
> +    cl.mode = "C";
> +    cl.module = get_modulename_from_path(location.file().as_string());
> +    cl.location = location;
> +    cl.type = t;
>
> I've tried to, but failed, since there does not seem to be any
> clang::NamedDecl (or any clang::Decl) available at this point.
>
> I've noticed several assumptions being made on the id string elsewhere;
> is it's format defined in ESBMC? For instance, the ***@***.***@***@***.***@z
> id above is purely generated by clang, yet there are many instances in
> which the id is compared to specific constants (mostly for intrinsics and
> functions in c2goto/library).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/711#discussion_r879259814>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHYDM2DHKHMH2GG37XLVLNIZ3ANCNFSM5WP7UZ6Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse committed:
[clang-c-frontend] clarify comment about compound literals in file scope

fbrausse committed:
[clang-c-frontend] fix compound literals' id to take the module name into account

fbrausse committed:
[clang-c-frontend] comment: clarify modelling storage for compound literals

fbrausse committed:
[clang-c-frontend] typos

fbrausse commented:
@mikhailramalho thanks for your inputs! Based on them I've clarified some parts via comments and assertions, and added 2 more regression tests. Could you take another look?

lucasccordeiro commented:
@fbrausse: have you considered the examples from this issue https://github.com/esbmc/esbmc/issues/709?

lucasccordeiro commented on regression/esbmc/github_591-1/main.c:
```diff
@@ -0,0 +1,5 @@
+#include <assert.h>
+int main()
+{
+	assert(((int){0} = 42) == 42);
```
Shall we use two spaces here?

fbrausse replied:
The regression suite isn't clang-formatted. I've changed the indentation in the 591-* cases.

lucasccordeiro commented on regression/esbmc/github_591-2/main.c:
```diff
@@ -0,0 +1,9 @@
+#include <string.h>
+int main()
+{
+	int z;
```
Something is wrong with the code indentation.

fbrausse replied:
Not really, it's the Linux kernel coding style ;) Anyway, changed.

fbrausse committed:
[regression] add 2 failing and 1 working tests from #709

The two failing tests are those that use "cast to union", which before the
recent introduction of new symbols for compound literals had been handled
the same way: by a constant_*2t expression for the construct.

Thus, the compound literals case works as expected now, but the cast to
union still fails.

fbrausse committed:
[regression] clang-format esbmc/github_591-*/*.c

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -466,7 +468,7 @@ bool clang_c_convertert::get_var(const clang::VarDecl &vd, exprt &new_expr)
   {
     // Initialize with zero value, if the symbol has initial value,
-    // it will be add later on this method
+    // it will be added later on in this method
```
Forgot to remove `on`?

fbrausse replied:
No, I saw "later on" as an idiom. Can remove "on", if you like.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1168,6 +1170,8 @@ bool clang_c_convertert::get_builtin_type(
   return false;
 }
 
+static std::string get_decl_name(const clang::NamedDecl &nd);
```
remove?

mikhailramalho commented on src/util/symbol_generator.h:
```diff
@@ -0,0 +1,23 @@
+
+#include <util/context.h>
+
+struct symbol_generator
```
can you make it a class?

fbrausse replied:
done

mikhailramalho commented on src/goto-programs/goto_sideeffects.cpp:
```diff
@@ -675,13 +675,13 @@ void goto_convertt::remove_function_call(
     new_base_name += '_';
     new_base_name += id2string(symbol.name);
-    new_base_name += "$" + std::to_string(++temporary_counter);
+    new_base_name += "$" + std::to_string(++tmp_symbol.counter);
```
do we need to manually increment the counter here? Shouldn't we be calling new_symbol?

fbrausse replied:
I didn't want to change semantics of code other than that touching compound literals. If we want to use new_symbol() here (which sounds like a good idea), we'd need to take a careful look at `goto_convertt::new_name()` and how to possibly merge it with `new_symbol()`. They are different: `new_symbol()` sets the `lvalue` attribute, which I'm not sure this code path here expects.

mikhailramalho commented on src/goto-programs/goto_sideeffects.cpp:
```diff
@@ -726,9 +726,9 @@ void goto_convertt::remove_cpp_new(
   symbolt new_symbol;
 
-  new_symbol.name = "new_ptr$" + std::to_string(++temporary_counter);
+  new_symbol.name = "new_ptr$" + std::to_string(++tmp_symbol.counter);
```
do we need to manually increment the counter here? Shouldn't we be calling new_symbol?

fbrausse replied:
See above.

fbrausse committed:
[clang-c-frontend] change signature of get_default_symbol() to avoid unnecessary conversions

lucasccordeiro commented:
@fbrausse: can I ask you to rebase this PR?

fbrausse commented:
> @fbrausse: can I ask you to rebase this PR?

Done.

rafaelsamenezes commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1444,7 +1446,54 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
+     * frequent copies of the symbol's name and id. */
+    std::string module_name =
+      get_modulename_from_path(location.file().as_string());
+    symbolt &cl = anon_symbol.new_symbol(context, t, module_name + "$cl$");
```
Wouldn't be better to be verbose here? I mean, I feel that in the CE the user will see this `$cl$` and will be lost.

fbrausse replied:
You are right.
```
State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main$cl$1.raw
```
I now changed it to:
```
State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main.c:14$compound-literal$1.raw
```

mikhailramalho replied:
we can also mark this assignment as hidden. but changing symex_assign, that way it won't appear on the cex unless we force printing the assignment.

Do we print this variable when creating a witness? It shouldn't be there.

fbrausse replied:
> we can also mark this assignment as hidden. but changing symex_assign, that way it won't appear on the cex unless we force printing the assignment.

Sorry, I don't understand, can you elaborate? Which assignment are you referring to (`var = ...` or the initialization of the compound-literal symbol?) and how can printing the assignment be forced?

fbrausse replied:
> Do we print this variable when creating a witness? It shouldn't be there.

This is the entire cex output for https://github.com/esbmc/esbmc/blob/b3ad1994b3d2f3cfb1dd9847b08a389368372703/regression/esbmc/github_709-cl/main.c
```
Counterexample:

State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main.c:14$compound-literal$1.raw

State 2 file main.c line 16 function main thread 0
----------------------------------------------------
  ta = 127 (00000000 00000000 00000000 01111111)

State 3 file main.c line 19 function main thread 0
----------------------------------------------------
Violated property:
  file main.c line 19 function main
  assertion ta == tb
  ta == tb


VERIFICATION FAILED
```

rafaelsamenezes replied:
@mikhailramalho meant when you use the `witness-output` option. Witness validators will not be able to handle the compound literal assignment. So we shouldn't show it and hope that the `ta = 127` is enough.

fbrausse replied:
@rafaelsamenezes Thanks, I wasn't aware.

@mikhailramalho this is the [witness](https://github.com/esbmc/esbmc/files/8765153/witness.txt) output. It does not contain the identifier nor the assignment (only `ta = 127;`), however, it does show
```xml
    <node id="N2"/>
    <edge id="E1" source="N1" target="N2">
      <data key="startline">14</data>
      <data key="threadId">0</data>
    </edge>
```
as the first edge in the graph corresponding to the assignment.

I have zero experience with these witness files; is it ok for this PR?

*edit* sorry, pasted wrong edge.

mikhailramalho replied:
Actually, the witness looks fine.

May I ask if you can do one last test? Can you initialize .raw and see what we get in the cex/witness?

fbrausse replied:
`int var = (s2){.raw = (&x)->data}.raw;` makes no difference.

`int var = ((s2){(&x)->data}.raw = 42);` obviously changes `ta = 42`, but that's it: [cex](https://github.com/esbmc/esbmc/files/8770055/cex.txt), [witness.txt](https://github.com/esbmc/esbmc/files/8770059/witness.txt)

`int var = ((s2){(&x)->data}.raw = (&x)->data);` again no change to the previous.

fbrausse replied:
If `var` becomes a struct type, the concrete value is printed, but the compound literal identifier is not: 
[test.c](https://github.com/esbmc/esbmc/files/8770086/test.c.txt), [cex.txt](https://github.com/esbmc/esbmc/files/8770090/cex.txt), [witness.txt](https://github.com/esbmc/esbmc/files/8770102/witness.txt)

mikhailramalho replied:
nice, so now my point is if we should print the:
```
State 1 file test.c line 14 function main thread 0
----------------------------------------------------
  var = test.c:14$compound-literal$1.raw
```

Can you hide it from the cex? I mean, unless there is a value, we don't usually print these assignments. WDYT @lucasccordeiro, @rafaelsamenezes?

fbrausse replied:
I find it rather confusing that the value of `var` is not given in the counter example. For me it would make more sense if the line read
```
  var = { .raw=127 }.raw
```
I mean, consider this shortened [example](https://github.com/esbmc/esbmc/files/8770278/test.c.txt)
```c
  int var = (s2){(&x)->data}.raw;
  int ta;
  ta = var;
  var = 0;
  assert(ta == var);
```
with nearly incomprehensible counter-example since `var`'s value is not shown:
```
Counterexample:

State 1 file test.c line 14 function main thread 0
----------------------------------------------------
  var = test.c:14$compound-literal$1.raw

State 2 file test.c line 16 function main thread 0
----------------------------------------------------
  ta = 127 (00000000 00000000 00000000 01111111)

State 3 file test.c line 18 function main thread 0
----------------------------------------------------
Violated property:
  file test.c line 18 function main
  assertion ta == var
  ta == var
```

fbrausse committed:
[clang-c-frontend] give symbols for compound literals a better name

mikhailramalho commented:
I think Lucas added an option that prints these internal symbols, even if
they are hidden, was it --compact-trace maybe?



Em ter., 24 de mai. de 2022 às 15:32, Franz Brauße ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/clang-c-frontend/clang_c_convert.cpp
> <https://github.com/esbmc/esbmc/pull/711#discussion_r880591429>:
>
> > +     * Therefore, in C, for instance a pointer to it can be taken and used
> +     * within the same block without restrictions, e.g.
> +     *
> +     *   int *p = &(int){0}, x = 42;
> +     *   memcpy(p, &x, sizeof(int));
> +     *   assert(*p == 42);
> +     *
> +     * It has the same semantics as an non-anonymous variable of the
> +     * corresponding type. Thus, we introduce a new symbol to represent it.
> +     */
> +
> +    /* "cl" stands for "compound literal", but that is too long for possibly
> +     * frequent copies of the symbol's name and id. */
> +    std::string module_name =
> +      get_modulename_from_path(location.file().as_string());
> +    symbolt &cl = anon_symbol.new_symbol(context, t, module_name + "$cl$");
>
> we can also mark this assignment as hidden. but changing symex_assign,
> that way it won't appear on the cex unless we force printing the assignment.
>
> Sorry, I don't understand, can you elaborate? Which assignment are you
> referring to (var = ... or the initialization of the compound-literal
> symbol?) and how can printing the assignment be forced?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/711#discussion_r880591429>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH5YP3PO3YSL5TFSAVTVLTSBRANCNFSM5WP7UZ6Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

mikhailramalho commented:
yeah, should we print this first assignment? I'd say we either print the
original code or hide it.

I think witness validators will have some problems with the '$'.


Em ter., 24 de mai. de 2022 às 15:36, Franz Brauße ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/clang-c-frontend/clang_c_convert.cpp
> <https://github.com/esbmc/esbmc/pull/711#discussion_r880596459>:
>
> > +     * Therefore, in C, for instance a pointer to it can be taken and used
> +     * within the same block without restrictions, e.g.
> +     *
> +     *   int *p = &(int){0}, x = 42;
> +     *   memcpy(p, &x, sizeof(int));
> +     *   assert(*p == 42);
> +     *
> +     * It has the same semantics as an non-anonymous variable of the
> +     * corresponding type. Thus, we introduce a new symbol to represent it.
> +     */
> +
> +    /* "cl" stands for "compound literal", but that is too long for possibly
> +     * frequent copies of the symbol's name and id. */
> +    std::string module_name =
> +      get_modulename_from_path(location.file().as_string());
> +    symbolt &cl = anon_symbol.new_symbol(context, t, module_name + "$cl$");
>
> Do we print this variable when creating a witness? It shouldn't be there.
>
> This is the entire cex output for
> https://github.com/esbmc/esbmc/blob/b3ad1994b3d2f3cfb1dd9847b08a389368372703/regression/esbmc/github_709-cl/main.c
>
> Counterexample:
>
> State 1 file main.c line 14 function main thread 0
> ----------------------------------------------------
>   var = main.c:14$compound-literal$1.raw
>
> State 2 file main.c line 16 function main thread 0
> ----------------------------------------------------
>   ta = 127 (00000000 00000000 00000000 01111111)
>
> State 3 file main.c line 19 function main thread 0
> ----------------------------------------------------
> Violated property:
>   file main.c line 19 function main
>   assertion ta == tb
>   ta == tb
>
>
> VERIFICATION FAILED
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/711#discussion_r880596459>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6BUR5EXLS7YEZGZV3VLTSOPANCNFSM5WP7UZ6Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

rafaelsamenezes commented on src/util/symbol_generator.h:
```diff
@@ -0,0 +1,24 @@
+
```
Could you add  a header like: https://github.com/esbmc/esbmc/blob/master/src/util/algorithms.h

fbrausse replied:
Strange to have copyright/authorship information only in the interface definition (header file). Who wrote the actual code I just moved to this separate class in order to not duplicate it, I don't know.

rafaelsamenezes replied:
I'm ok with skipping authorship. But it is still good to know why you created a new header file, it makes easier to use/extend later.

fbrausse replied:
done

rafaelsamenezes commented on src/util/symbol_generator.h:
```diff
@@ -0,0 +1,24 @@
+  {
+  }
+
+  symbolt &new_symbol(
```
Could you write some doxygen comments?

fbrausse replied:
done

rafaelsamenezes commented on src/util/symbol_generator.cpp:
```diff
@@ -0,0 +1,22 @@
+    new_symbol.type = type;
+  } while(context.move(new_symbol, symbol_ptr));
+
+  return *symbol_ptr;
```
Is there any chance for this to cause a segmentation fault by `symbol_ptr` being `NULL`? Could you just guard it with an assertion?

fbrausse replied:
There is not, see https://github.com/esbmc/esbmc/blob/35f03906d83bb9ae56d9f97223142e10f54bf287/src/util/context.cpp#L26

rafaelsamenezes replied:
Ow nice

fbrausse committed:
[util] symbol_generator: add some docs
